core/trino-main/src/test/java/io/trino/server/security/oauth2/TestingHydraIdentityProvider.java

```diff
@@ -13,76 +13,103 @@
+import static javax.servlet.http.HttpServletResponse.SC_NOT_FOUND;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+import static org.testng.Assert.assertNotNull;
 
 public class TestingHydraIdentityProvider
         implements AutoCloseable
 {
-    static final int TTL_ACCESS_TOKEN_IN_SECONDS = 30;
     private static final String HYDRA_IMAGE = "oryd/hydra:v1.9.0-sqlite";
 
+    private final Duration ttlAccessToken;
     private final Network network = Network.newNetwork();
-
-    private final FixedHostPortGenericContainer<?> consentContainer = new FixedHostPortGenericContainer<>("oryd/hydra-login-consent-node:v1.4.2")
-            .withNetwork(network)
-            .withNetworkAliases("consent")
-            .withExposedPorts(3000)
-            .withEnv("HYDRA_ADMIN_URL", "https://hydra:4445")
-            .withEnv("NODE_TLS_REJECT_UNAUTHORIZED", "0")
-            .waitingFor(Wait.forHttp("/").forStatusCode(200));
-
-    private final FixedHostPortGenericContainer<?> hydraContainer = createHydraContainer()
-            .withNetworkAliases("hydra")
-            .withExposedPorts(4444, 4445)
-            .withEnv("DSN", "memory")
-            .withEnv("URLS_SELF_ISSUER", "https://hydra:4444/")
-            .withEnv("URLS_CONSENT", "http://consent:3000/consent")
-            .withEnv("URLS_LOGIN", "http://consent:3000/login")
-            .withEnv("SERVE_TLS_KEY_PATH", "/tmp/certs/localhost.pem")
-            .withEnv("SERVE_TLS_CERT_PATH", "/tmp/certs/localhost.pem")
-            .withEnv("STRATEGIES_ACCESS_TOKEN", "jwt")
-            .withEnv("TTL_ACCESS_TOKEN", TTL_ACCESS_TOKEN_IN_SECONDS + "s")
-            .withCommand("serve", "all")
-            .withCopyFileToContainer(MountableFile.forClasspathResource("/cert"), "/tmp/certs")
-            .waitingFor(new WaitAllStrategy()
-                    .withStrategy(Wait.forLogMessage(".*Setting up http server on :4444.*", 1))
-                    .withStrategy(Wait.forLogMessage(".*Setting up http server on :4445.*", 1)));
-
     private final AutoCloseableCloser closer = AutoCloseableCloser.create();
+    private TestingHttpServer loginAndConsentServer;
+    private GenericContainer<?> hydraContainer;
+
+    public TestingHydraIdentityProvider()
+    {
+        this(Duration.ofMinutes(30));
+    }
 
-    TestingHydraIdentityProvider()
+    public TestingHydraIdentityProvider(Duration ttlAccessToken)
     {
+        this.ttlAccessToken = requireNonNull(ttlAccessToken, "ttlAccessToken is null");
         closer.register(network);
-        closer.register(consentContainer);
-        closer.register(hydraContainer);
     }
 
     public void start()
+            throws Exception
     {
-        consentContainer.start();
+        loginAndConsentServer = createTestingLoginAndConsentServer();
+        closer.register(loginAndConsentServer::stop);
+        loginAndConsentServer.start();
+        URI loginAndConsentBaseUrl = loginAndConsentServer.getBaseUrl();
```

Comment:
How does it work? It looks like you are accessing host port from docker. Is it just because http server is not bound on localhost?

Reply:
There's no communication hydra -> login & consent service, only login & consent service -> hydra, so it's not a problem that hydra cannot access a service on the host.
That said, hydra needs login & consent urls configured because it redirects the client to those urls to log in and/or give consent so the client needs access.